conda_package_streaming/conda_fmt.py

```diff
@@ -0,0 +1,179 @@
+
+import json
+import shutil
+import tarfile
+import tempfile
+import zipfile
+from contextlib import contextmanager
+from pathlib import Path
+from typing import Callable, Iterator
+
+import zstandard
+
+# increase to reduce speed and increase compression (levels above 19 use much
+# more memory)
+ZSTD_COMPRESS_LEVEL = 19
+# increase to reduce compression and increase speed
+ZSTD_COMPRESS_THREADS = 1
+
+CONDA_PACKAGE_FORMAT_VERSION = 2
+
+# Account for growth from "2 GB of /dev/urandom" to not exceed ZIP64_LIMIT after
+# compression
+CONDA_ZIP64_LIMIT = zipfile.ZIP64_LIMIT - (1 << 18) - 1
+
+
+def anonymize(tarinfo: tarfile.TarInfo):
+    """
+    Pass to ``tarfile.add(..., filter=anonymize)`` to anonymize uid/gid.
+
+    Does not anonymize mtime or any other field.
+    """
+    tarinfo.uid = tarinfo.gid = 0
+    tarinfo.uname = tarinfo.gname = ""
+    return tarinfo
+
+
+class CondaTarFile(tarfile.TarFile):
+    """
+    Subclass of :external+python:py:class:`tarfile.TarFile` that adds members to
+    a second ``info`` tar if they match ``is_info(name)``.
+
+    Create this with ``conda_builder(...)`` which sets up the component
+    archives, then wraps them into a ``.conda`` on exit.
+
+    Only useful for creating, not extracting ``.conda``.
+    """
+
+    info_tar: tarfile.TarFile
+    is_info: Callable
+
+    def __init__(
+        self,
+        *args,
+        info_tar: tarfile.TarFile,
+        is_info=lambda name: name.startswith("info/"),
+        **kwargs,
+    ):
+        super().__init__(*args, **kwargs)
+        self.info_tar = info_tar
+        self.is_info = is_info
+
+    def addfile(self, tarinfo, fileobj=None):
+        """
+        Add the TarInfo object ``tarinfo`` to the archive. If ``fileobj`` is
+        given, it should be a binary file, and tarinfo.size bytes are read from
+        it and added to the archive. You can create TarInfo objects directly, or
+        by using ``gettarinfo()``.
+
+        If ``self.is_info(tarinfo.name)`` returns ``True``, add ``tarinfo`` to
+        ``self.info_tar`` instead.
+        """
+        if self.is_info(tarinfo.name):
+            return self.info_tar.addfile(tarinfo, fileobj=fileobj)
+        else:
+            return super().addfile(tarinfo, fileobj)
+
+
+@contextmanager
+def conda_builder(
+    file_id,
+    path,
+    *,
+    compressor: Callable[
+        [], zstandard.ZstdCompressor
+    ] = lambda: zstandard.ZstdCompressor(
+        level=ZSTD_COMPRESS_LEVEL, threads=ZSTD_COMPRESS_THREADS
+    ),
+    is_info: Callable[[str], bool] = lambda filename: filename.startswith("info/"),
+    encoding="utf-8",
+) -> Iterator[CondaTarFile]:
+    """
+    Produce a ``TarFile`` subclass used to build a ``.conda`` package. The
+    subclass delegates ``addfile()`` to the ``info-`` component when ``is_info``
+    returns True.
+
+    When the context manager exits, ``{path}/{file_id}.conda`` is written with
+    the component tar archives.
+
+    Args:
```

Comment:
Docsting format enabled by sphinx.ext.napoleon

Reply:
Sphinx will automatically use type annotations when rendering API output, does this work with napoleon (can't see a preview in the PR)?

Reply:
Yes it works beautifully. https://www.sphinx-doc.org/en/master/usage/extensions/napoleon.html#docstrings